pallets/crowdloan-claim/src/lib.rs

```diff
@@ -704,35 +707,52 @@ pub mod pallet {
 					if !ProcessedClaims::<T>::contains_key((
 						&relaychain_account_id,
 						Self::curr_index(),
-					)) && Self::verify_contributor_identity_proof(
-						relaychain_account_id.clone(),
-						parachain_account_id.clone(),
-						identity_proof.clone(),
-					)
-					.is_ok() && Self::verify_contribution_proof(contribution_proof.clone())
+					)) {
+						if Self::verify_contributor_identity_proof(
+							relaychain_account_id.clone(),
+							parachain_account_id.clone(),
+							identity_proof.clone(),
+						)
 						.is_ok()
-					{
-						// Only the claim reward transaction can be invoked via an unsigned regime
-						return ValidTransaction::with_tag_prefix("CrowdloanClaimReward")
-							.priority(T::ClaimTransactionPriority::get())
-							.and_provides((
-								relaychain_account_id,
-								parachain_account_id,
-								identity_proof,
-								contribution_proof,
-								contribution,
-							))
-							.longevity(
-								TryInto::<u64>::try_into(T::ClaimTransactionLongevity::get())
-									.unwrap_or(64_u64),
+						{
+							let mut leaf_data = relaychain_account_id.encode();
+							leaf_data.extend_from_slice(&contribution.encode());
+
+							if Self::verify_contribution_proof(
+								contribution_proof.clone(),
+								&leaf_data,
 							)
-							.propagate(true)
-							.build();
+							.is_ok()
+							{
+								// Only the claim reward transaction can be invoked via an unsigned regime
+								return ValidTransaction::with_tag_prefix("CrowdloanClaimReward")
+									.priority(T::ClaimTransactionPriority::get())
+									.and_provides((
+										relaychain_account_id,
+										parachain_account_id,
+										identity_proof,
+										contribution_proof,
+										contribution,
+									))
+									.longevity(
+										TryInto::<u64>::try_into(
+											T::ClaimTransactionLongevity::get(),
+										)
+										.unwrap_or(64_u64),
+									)
+									.propagate(true)
+									.build();
+							} else {
+								return InvalidTransaction::Custom(3).into();
+							}
+						} else {
+							return InvalidTransaction::Custom(2).into();
```

Comment:
I wouldn't give more hints to a potential attacker of where the code failed, I would stick to BadOrigin for all.